fleetctl/fleetctl.go

```diff
@@ -1140,64 +1132,37 @@ func checkSystemdActiveState(apiStates []*schema.UnitState, name string, maxAtte
 	}
 }
 
-func assertFetchSystemdActiveState(apiStates []*schema.UnitState, name string) error {
-	if err := assertSystemdActiveState(apiStates, name); err == nil {
-		return nil
-	}
-
-	// If the assertion failed, we again need to get unit states via cAPI,
-	// to retry the assertion repeatedly.
-	//
-	// NOTE: Ideally we should be able to fetch the state only for a single
-	// unit. However, we cannot do that for now, because cAPI.UnitState()
-	// is not available. In the future we need to implement cAPI.UnitState()
-	// and all dependendent parts all over the tree in fleet, (schema,
-	// etcdRegistry, rpcRegistry, etc) to replace UnitStates() in this place
-	// with the new method UnitState(). In practice, calling UnitStates() here
-	// is not as badly inefficient as it looks, because it will be anyway
-	// rarely called only when the assertion failed. - dpark 20160907
-
-	time.Sleep(defaultSleepTime)
+// assertSystemdActiveState determines if a given systemd unit is actually
+// in the active state, making use of cAPI.
+// It repeatedly checks up to defaultSleepTimeout. If ActiveState of the given
+// unit is active and LoadState of the given unit is loaded.
+// If it cannot get the expected states within the period, return error.
+func assertSystemdActiveState(unitName string) error {
+	var errbuf error
+	fetchSystemdActiveState := func() bool {
```

Comment:
Why not just have this `func() err` and propagate the error instead of using this buffer variable?